eclipse-che/che-dashboard
# Dashboard frontend sanity check

### What does this PR do?

This PR adds the custom redux middleware to the dashboard. The middleware triggers the sanity check each time before requesting either the che-server API or the dashboard backend. The sanity check results get cached, so no unnecessary extra requests will be done.
If the sanity check fails, only one error alert will be shown. All other error alerts will be suppressed.

#### Screenshots

<details>
<summary>In preload phase</summary>

<img width="890" alt="Screenshot 2022-11-28 at 17 00 31" src="https://user-images.githubusercontent.com/16220722/204490828-0c845715-a03a-43aa-b326-e8190a60529e.png">

</details>

<details>
<summary>When the dashboard loaded</summary>

<img width="891" alt="Screenshot 2022-11-29 at 11 21 54" src="https://user-images.githubusercontent.com/16220722/204490808-f4077e2f-4afc-4c20-8b06-78da60b95e24.png">
</details>



### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/21799
	- add API sanity check with a single request to prevent showing multiple error messages
	- if after all checks above and re-tries API is still failing we show a single error with the session expired message and redirect to the login page

### Is it tested? How?


1. Open the dashboard in two windows simultaneously and sign in: https://devspaces.apps.cluster-7q8gj.7q8gj.sandbox838.opentlc.com/
2. To test this fix in preload phase: 
	- Press Ctrl(Cmd)+R in the first window and immediately sign out in the second window. 
	- You should see the error page with the corresponding message;
3. To test the fix in the running application
	- Wait at least 30 seconds after the dashboard is loaded;
	- Sign out in one of the opened windows;
	- Open the workspaces list page in the second window;
	- You should see the error alert notification.
